### PR TITLE
refactor: move `Constant` from `Ast` to `shared`

### PR DESCRIPTION
**Pull request by sockmaster27**
Builds on #8577 
Related to #7871

### CODE REVIEW

**Comment by magnus-madsen** [comment]
Some CI errors

**Comment by sockmaster27** [comment]
😵‍💫

File: main/src/ca/uwaterloo/flix/language/ast/OccurrenceAst.scala [hunk 1]
```diff
@@ -45,7 +45,7 @@ object OccurrenceAst {
 
   object Expr {
 
-    case class Constant(cst: Ast.Constant, tpe: MonoType, loc: SourceLocation) extends OccurrenceAst.Expr {
+    case class Constant(cst: shared.Constant, tpe: MonoType, loc: SourceLocation) extends OccurrenceAst.Expr {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/OccurrenceAst.scala:48]: Just import directly

File: main/src/ca/uwaterloo/flix/language/phase/Safety.scala [hunk 2]
```diff
@@ -221,8 +221,6 @@ object Safety {
     defn.spec.eff match {
       case Type.Pure => true
       case Type.IO => true
-      case Type.NonDet => true
-      case Type.Sys => true
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Safety.scala:225]: What happened here in this PR?

File: main/src/ca/uwaterloo/flix/language/ast/OccurrenceAst.scala (hunk 1 above)
Comment (sockmaster27) [reply main/src/ca/uwaterloo/flix/language/ast/OccurrenceAst.scala:48]: The name clashes though

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/OccurrenceAst.scala:48]: Ah, I think Expr.Constant should be Expr.Cst

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/OccurrenceAst.scala:48]: Solving the clash.

**Review by magnus-madsen — APPROVED** [review]

**Pull request merged**